homeassistant/components/withings/common.py

```diff
@@ -335,3 +338,13 @@ def get_data_manager(
     )
 
     return dm_dict[entry_id]
+
+
+class WithingsLocalOAuth2Implementation(LocalOAuth2Implementation):
+    """Oauth2 implementation that only uses the external url."""
+
+    @property
+    def redirect_uri(self) -> str:
+        """Return the redirect uri."""
+        url = get_url(self.hass, prefer_cloud=True)
```

Comment:
This still prefers internal above external, as cloud is an external URL. However, if an external URL is used, it will prefer the cloud one over the manual set one.

I guess you are looking for this:

```suggestion
        url = get_url(self.hass, prefer_external=True, prefer_cloud=True)
```

Resulting in: Give preferably an external URL, and if possible preferably a Cloud URL, if none available, fallback to internal URL logic (either user set or detected IP on the network interface).

Reply:
Pretty close. I pushed another commit that disables internal url and prefers both the external and cloud urls.